components/logins/src/logins.udl

```diff
@@ -0,0 +1,116 @@
+namespace logins {
+
+    [Throws=LoginsError]
+    string open_and_get_salt(string path, [ByRef] string encryption_key);
+
+    [Throws=LoginsError]
+    void open_and_migrate_to_plaintext_header(string path, [ByRef] string encryption_key, [ByRef] string salt);
+};
+
+
+// Converted from mozilla.appservices.logins.protobuf.rs
+dictionary LoginRecord {
+    string guid;
+    string hostname;
+    string password;
+    string username;
+    string? http_realm;
+    string? form_submit_url;
+    string username_field;
+    string password_field;
+    i64 times_used;
+    i64 time_created;
+    i64 time_last_used;
+    i64 time_password_changed;
+};
+
+dictionary MigrationMetrics {
+    MigrationPhaseMetrics fixup_phase;
+    MigrationPhaseMetrics insert_phase;
+    u64 num_processed;
+    u64 num_succeeded;
+    u64 num_failed;
+    u64 total_duration;
+    sequence<string> errors;
+};
+
+dictionary MigrationPhaseMetrics {
+    u64 num_processed;
+    u64 num_succeeded;
+    u64 num_failed;
+    u64 total_duration;
+    sequence<string> errors;
+};
+
+[Error]
+enum LoginsError {
+    "Unexpected",
+    "SyncAuthInvalid",
+    "MismatchedLock",
+    "NoSuchRecord",
+    "IdCollision",
+    "InvalidRecord",
+    "InvalidKey",
+    "RequestFailed",
+    "Interrupted"
+};
+
+interface PasswordStore {
+    [Throws=LoginsError]
+    constructor(string path, [ByRef] string encryption_key);
+
+    [Name=new_with_salt, Throws=LoginsError]
+    constructor(string path, [ByRef] string encryption_key, [ByRef] string salt);
+
+    [Throws=LoginsError]
+    void check_valid_with_no_dupes(LoginRecord record);
+
+    [Throws=LoginsError]
+    string add(LoginRecord record);
+
+    [Throws=LoginsError]
+    boolean delete([ByRef] string id);
+
+    [Throws=LoginsError]
+    void wipe();
+    [Throws=LoginsError]
+    void wipe_local();
+    [Throws=LoginsError]
+    void reset();
+
+    [Throws=LoginsError]
+    void disable_mem_security();
+
+    [Throws=LoginsError]
+    void rekey_database([ByRef] string new_encryption_key);
+
+    [Throws=LoginsError]
+    void touch([ByRef] string id);
+
+    [Throws=LoginsError]
+    sequence<LoginRecord> list();
+
+    [Throws=LoginsError]
+    sequence<LoginRecord> get_by_base_domain([ByRef] string base_domain);
+
+    [Throws=LoginsError]
+    sequence<LoginRecord> potential_dupes_ignoring_username(LoginRecord record);
+
+    [Throws=LoginsError]
+    LoginRecord? get([ByRef] string id);
+
+    [Throws=LoginsError]
+    void update(LoginRecord record);
+
+    [Throws=LoginsError]
+    MigrationMetrics import_multiple(sequence<LoginRecord> record);
+
+    // Methods NOT ported over from ffi/lib.rs
+
+    // Q1: sync15_passwords_sync(), do we still need to call sync in stores.rs?
```

Comment:
So I was able to delete the ffi crate completely! This and the question below were the only ones that didn't make it over to uniffi. It would simplify things a lot if we don't have to port over these two

Reply:
I don't see references for either of these functions in firefox ios, android components, fenix, or mozilla central so let's leave them out for now.